### ISSUE TITLE
NumberProperty: add support for validValues and isValidValue

### ISSUE THREAD

**Issue by zepumph** [phetsims/axon#153]
as part of https://github.com/phetsims/phet-io/issues/1236 @samreid and I found that for `Sim.screenIndexProperty` we were forced into a using a `Range` of possible screen indices rather than value values, which seemed a bit better for the instance proxies UI to interpret. We saw that @pixelzoom added in these asserts back when he authored the file. @pixelzoom do you have an objection to supporting valid values in NumberProperty?

**Comment by pixelzoom** [comment]
`NumberProperty` needs validators for both `Range` and an array of values.

**Comment by samreid** [comment]
Implemented in the preceding commit, @zepumph can you please review?

**Comment by pixelzoom** [comment]
Problem with the above commit - you didn't verify that validValues is an array of numbers.

**Comment by pixelzoom** [comment]
E.g. this is currently allowed:

```js
> var p = new phet.axon.NumberProperty( 'foo', { validValues: [ 'foo', 'bar' ] } )
```

**Comment by pixelzoom** [comment]
Something like this (untested) is needed:

```js
assert && assert( _.every( options.validValues, function( value ) { return IS_NUMBER( value ); } );
```

**Comment by pixelzoom** [comment]
In retrospect, perhaps this change wasn't really needed. When `validValues` is provided, all it's doing is verifying that they are numbers, then all other responsibility is delegated to Property.  And when you know the set of valid values, do you really need a `NumberProperty` (or any type-specific Property for that matter)?   E.g. you could have used a plain old `Property` for Sim `screenIndexProperty`, and it would have worked just fine.  Perhaps there's some addition benefit gained from having `TNumberProperty` involved that I'm not aware of.

**Comment by zepumph** [comment]
NumberProperty has extra metadata provided over the phet-io api which helps in the wrapper.

**Comment by pixelzoom** [comment]
While we're at it...  It seems like being able to provide `isValidValue` is also important.  For example, in equality-explorer, I need a NumberProperty whose values is an integer >= 0.  There's no way that I can do that with the current interface, so I can either (a) use a NumberProperty and blow off validation, (b) use a NumberProperty and add a separate listener to validate, or (c) use a Property and defer the pain until PhET-iO instrumentation.  I'm currently leaning towards (b), but then there's really no non-PhET-iO benefit to using a NumberProperty.

**Comment by pixelzoom** [comment]
I also see this problem:

```js
92 valueType: 'FloatingPoint', // 'FloatingPoint' | 'Integer'
110 this.valueType = options.valueType;
126 this.valueType = options.valueType;
```

Duplicate assignment to `this.valueType`, no documentation that it's used by TNumberProperty, and there's nothing that verifies that the value actually matches valueType.

**Comment by zepumph** [comment]
Duplication and documentation fixed in https://github.com/phetsims/axon/commit/03294044fe624e03be296eab5842a3e5921d1d65

**Comment by pixelzoom** [comment]
I'm going to jump in and add support for `isValidValue`, since I need it in equality-explorer. I notified @zepumph via Slack.

**Comment by pixelzoom** [comment]
Support for `isValidValue` added in the above commit, along with some other cleanup. Note that I did _not_ address the issue of using `valueType` in the validation.

Console tests:

```
// default validation
> var p = new phet.axon.NumberProperty( 'foo' )
Assertion failed: invalid initial value: foo
> var p = new phet.axon.NumberProperty( 0 )
undefined
> p.value = 1
1
> p.value = 'foo'
Assertion failed: invalid value: foo

// range
> var p = new phet.axon.NumberProperty( 0, {
  range: { min:0, max:10 }
} )
undefined
> p.value = 5
5
> p.value = 11
Assertion failed: invalid value: 11
> p.value = -1
Assertion failed: invalid value: -1

// validValues
var p = new phet.axon.NumberProperty( 0, {
  validValues: [ 0, 1, 2 ]
} )
> p.value = 1
1
> p.value = 2
2
> p.value = 3
Assertion failed: invalid value: 3

// isValidValue
> var p = new phet.axon.NumberProperty( 0, {
  isValidValue: function( value ) { return value >= 0; }
} )
undefined
> p.value = 1
1
> p.value = 0
0
> p.value = -1
Assertion failed: invalid value: -1
> p.value = 'foo'
Assertion failed: invalid value: foo

// mutually-exclusive options
> var p = new phet.axon.NumberProperty( 0, {
  range: { min:0, max:10 },
  isValidValue: function( value ) { return value >= 0; },
  validValues: [ 0, 1, 2 ]
} )
undefined
Assertion failed: validValues, isValidValue and range are mutually-exclusive options
```

 The above tests should be added to axon units tests but (like https://github.com/phetsims/axon/issues/156#issuecomment-339147946) I was unsure on how to add them.

@zepumph please review.

**Comment by pixelzoom** [comment]
Re this TODO in NumberProperty:

```js
// valid values for options.units // TODO: Should this be an enum?  @zepumph says probably
var VALID_UNITS = [
```

I think they are fine as strings. This is how we typically handle options that has a set of valid string values.

**Comment by samreid** [comment]
@zepumph as part of your review can you please convert https://github.com/phetsims/axon/issues/153#issuecomment-339158399 to unit tests as was done in https://github.com/phetsims/axon/commit/c6602e547d9a609f4b706416293c29a0477e04c6

**Comment by zepumph** [comment]
Unit tests added and all are passing. Closing

**Comment by pixelzoom** [comment]
https://github.com/phetsims/axon/issues/153#issuecomment-339158399:
>  Note that I did _not_ address the issue of using valueType in the validation.

So we're not done yet. Reopening.

**Comment by zepumph** [comment]
Right! I'll do that part, thanks.

**Comment by pixelzoom** [comment]
Looking at the [continuous testing report](https://bayes.colorado.edu/continuous-testing/aqua/html/continuous-report.html), it looks like axon unit tests for NumberProperty might be causing problems in kite and scenery.  At 10/30/2107 4:22:49pm:

kite:
```
kite : unit-tests : require.js
1800 out of 1801 tests passed. 1 failed.
Axon: Simple Tests: Test NumberProperty failed:
Died on test #3     at https://bayes.colorado.edu/continuous-testing/snapshot-1509402169463/axon/tests/qunit/js/simple-tests.js?random=0.4610834488:356:3
    at https://bayes.colorado.edu/continuous-testing/snapshot-1509402169463/axon/tests/qunit/js/simple-tests.js?random=0.4610834488:416:3: phet is not defined
ReferenceError: phet is not defined
    at Object.<anonymous> (https://bayes.colorado.edu/continuous-testing/snapshot-1509402169463/axon/tests/qunit/js/simple-tests.js?random=0.4610834488:371:5)
```

scenery:
```
kite : unit-tests : require.js
1800 out of 1801 tests passed. 1 failed.
Axon: Simple Tests: Test NumberProperty failed:
Died on test #3     at https://bayes.colorado.edu/continuous-testing/snapshot-1509402169463/axon/tests/qunit/js/simple-tests.js?random=0.4610834488:356:3
    at https://bayes.colorado.edu/continuous-testing/snapshot-1509402169463/axon/tests/qunit/js/simple-tests.js?random=0.4610834488:416:3: phet is not defined
ReferenceError: phet is not defined
    at Object.<anonymous> (https://bayes.colorado.edu/continuous-testing/snapshot-1509402169463/axon/tests/qunit/js/simple-tests.js?random=0.4610834488:371:5)
```

**Comment by pixelzoom** [comment]
Similar NumberProperty test failure for dot at 10/30/2017 4:42:08

**Comment by pixelzoom** [comment]
Looks like unit tests were failing because they use `phet.axon.NumberProperty`. In https://github.com/phetsims/axon/commit/abbfd3073e4b5d3ada214fef8c085ee51a4ba30f, @jonathanolson recently removed the `phet.` prefix for other axon types, e.g. `axon.Property`.

**Comment by zepumph** [comment]
@pixelzoom does everything else look good for this one?

**Comment by zepumph** [comment]
I added validation for the valueType option as well as unit tests for them. Could you review the changes @pixelzoom?

**Comment by pixelzoom** [comment]
Issues:

(1) `isNumber` is private, used only in 1 place, and trivial, so I've inlined it.

(2) NumberProperty is using `Number.isInteger`, which was introduced in ES6, and is not supported in IE, [specification here])US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger). I tried changing to DOT.Util.isInteger, but that causes unit tests to fail -- axon can't depend on dot, because dot depends on axon.  Back to @zepumph to address this.
